libminifi/src/c2/ControllerSocketProtocol.cpp

```diff
@@ -0,0 +1,245 @@
+        return;
+      }
+      switch (head) {
+        case Operation::START:
+        {
+          std::string componentStr;
+          int size = stream.readUTF(componentStr);
+          if ( size != -1 ) {
+            auto components = update_sink_->getComponents(componentStr);
+            for (auto component : components) {
+              component->start();
+            }
+          } else {
+            logger_->log_debug("Connection broke with fd %d", fd);
+          }
+        }
+        break;
+        case Operation::STOP:
+        {
+          std::string componentStr;
+          int size = stream.readUTF(componentStr);
+          if ( size != -1 ) {
+            auto components = update_sink_->getComponents(componentStr);
+            for (auto component : components) {
+              component->stop(true, 1000);
+            }
+          } else {
+            logger_->log_debug("Connection broke with fd %d", fd);
+          }
+        }
+        break;
+        case Operation::CLEAR:
+        {
+          std::string connection;
+          int size = stream.readUTF(connection);
+          if ( size != -1 ) {
+            update_sink_->clearConnection(connection);
+          }
+        }
+        break;
+        case Operation::UPDATE:
+        {
+          std::string what;
+          int size = stream.readUTF(what);
+          if (size == -1) {
+            logger_->log_debug("Connection broke with fd %d", fd);
+            break;
+          }
+          if (what == "flow") {
+            std::string ff_loc;
+            int size = stream.readUTF(ff_loc);
+            std::ifstream tf(ff_loc);
+            std::string configuration((std::istreambuf_iterator<char>(tf)),
+                std::istreambuf_iterator<char>());
+            if (size == -1) {
+              logger_->log_debug("Connection broke with fd %d", fd);
+              break;
+            }
+            update_sink_->applyUpdate(configuration);
+          }
+        }
+        break;
+        case Operation::DESCRIBE:
+        {
+          std::string what;
+          int size = stream.readUTF(what);
+          if (size == -1) {
+            logger_->log_debug("Connection broke with fd %d", fd);
+            break;
+          }
+          if (what == "queue") {
+            std::string connection;
+            int size = stream.readUTF(connection);
+            if (size == -1) {
+              logger_->log_debug("Connection broke with fd %d", fd);
+              break;
+            }
+            std::stringstream response;
+            {
+              std::lock_guard<std::mutex> lock(controller_mutex_);
+              response << queue_size_[connection] << " / " << queue_max_[connection];
+            }
+            io::BaseStream resp;
+            resp.writeData(&head, 1);
+            resp.writeUTF(response.str());
+            write(fd, resp.getBuffer(), resp.getSize());
+          } else if (what == "processors") {
```

Comment:
minor:  processors is a little broad given that we are getting all components so this is bundling in ports as well.  does components make more sense?